src/jdk.httpserver/share/classes/com/sun/net/httpserver/HttpExchange.java

```diff
@@ -199,9 +210,46 @@ protected HttpExchange() {
      *                       specified and no response body may be written.
      * @throws IOException   if the response headers have already been sent or an I/O error occurs
      * @see   HttpExchange#getResponseBody()
+     * @see   HttpExchange#CHUNKED_CONTENT
+     * @see   HttpExchange#NO_CONTENT
      */
     public abstract void sendResponseHeaders(int rCode, long responseLength) throws IOException;
 
+    /**
+     * convenience method to send a "no content response"
+     * @param code the response code to send
+     * @throws IOException if the response headers have already been sent or an I/O error occurs
+     * @see HttpExchange#sendResponseHeaders(int, long)
+     */
+    public final void sendResponseHeadersNoContent(int code) throws IOException {
+        sendResponseHeaders(code,NO_CONTENT);
+    }
+    /**
+     * convenience method to send a chunked response. the caller must close the returned output stream.
+     * @param code the response code to send
+     * @throws IOException if the response headers have already been sent or an I/O error occurs
+     * @return the stream to write the response to. the caller must close the stream.
+     * @see HttpExchange#sendResponseHeaders(int, long)
+     */
+    public final OutputStream sendResponseHeadersChunked(int code) throws IOException {
+        sendResponseHeaders(code,CHUNKED_CONTENT);
+        return getResponseBody();
+    }
+    /**
+     * convenience method to send a fixed length response. the output stream is automatically closed and no
+     * further writes are permitted.
+     * @param code the response code to send
+     * @param data the data to send
+     * @throws IOException if the response headers have already been sent or an I/O error occurs
+     * @see HttpExchange#sendResponseHeaders(int, long)
+     */
+    public final void sendResponseHeaders(int code,byte[] data) throws IOException {
```

Comment:
I'd prefer to drop this method. Some time in the future we might come up with a better mechanism for handling request and response bodies and translating them to higher level types.

Reply:
It seems like this is a common source of issues though. If the provided length does not match the actual length of the content sent, the connection is completely borked. Clearly this isn't suitable for large responses, and the chunked should be used in that case, so maybe a note around this?

Even if the higher level types became available it would seem trivial to map this to those - similar to how the BodyHandlers work in HttpClient - common ones are provided - this would use those in reverse.

I don't think it is something to hold up the other changes, but I think it warrants a bit more discussion.

Reply:
If we keep this method then I think it should be named `sendResponse` since it sends the response (and not only the headers). It seems a bit limiting that it requires a single byte array - but on the other hand that's what we use in probably 80% of the tests. On the other hand it seems to promote a single `byte[]` array as first class citizen - and as @Michael-Mc-Mahon  says, that might not be the best choice. Agreed that we can discuss it though.

Reply:
I think sendResponse() make more sense to me. We could also add a OutputStream sendResponse(int code,int length), and/or add OutputStream sendResponseChunked(int code) and try to break the dependency on the sendResponseHeaders obtuse length parameter.

In reality though, it is either going to be a simple byte array/string, or it is probably best sent using an output stream - and using a fixed length stream for that rather than chunked is inefficient because you need to buffer the entire response (unless you know it ahead of time like in a file transfer) - so that should be avoided.